src/core/deltafilewrapper.cpp

```diff
@@ -830,14 +830,52 @@ void DeltaFileWrapper::mergeCreateDelta( const QJsonObject &delta )
 {
   Q_ASSERT( delta.value( QStringLiteral( "method" ) ) == "create" );
 
+  QJsonObject newData = delta.value( QStringLiteral( "new" ) ).toObject();
+
   const QString localPk = delta.value( QStringLiteral( "localPk" ) ).toString();
   const QString localLayerId = delta.value( QStringLiteral( "localLayerId" ) ).toString();
-  mLocalPkToDeltaUuid[localLayerId][localPk] = delta.value( QStringLiteral( "uuid" ) ).toString();
+  const QString existingDeltaUuid = mLocalPkToDeltaUuid[localLayerId].take( localPk );
+
+  if ( !existingDeltaUuid.isEmpty() )
+  {
+    const int existingDeltaIdx = getDeltaIndexByUuid( existingDeltaUuid );
+
+    Q_ASSERT( existingDeltaIdx >= 0 );
+
+    const QJsonObject existingDelta = mDeltas.at( existingDeltaIdx ).toObject();
+    const QString existingDeltaMethod = existingDelta.value( QStringLiteral( "method" ) ).toString();
+
+    // There is a change that the current "create" delta is actually the "undo" of a "delete" delta from earlier.
+    // In those cases we just discard both deltas like nothing happened
+    if ( existingDeltaMethod == QStringLiteral( "delete" ) )
+    {
+      const QJsonObject existingDeltaOldData = existingDelta.value( QStringLiteral( "old" ) ).toObject();
+
+      // the newly added "create" delta matches 1:1 with previously existing "delete" delta
+      if ( existingDeltaOldData == newData )
+      {
+        Q_ASSERT( existingDelta.value( QStringLiteral( "localLayerId" ) ).toString() == localLayerId );
+        Q_ASSERT( existingDelta.value( QStringLiteral( "localPk" ) ).toString() == localPk );
+
+        mDeltas.removeAt( existingDeltaIdx );
+        mIsDirty = true;
+
+        emit countChanged();
+
+        qInfo() << "DeltaFileWrapper::mergeCreateDelta: removed delete delta instead of adding a create delta: " << existingDeltaIdx;
+
+        return;
+      }
+    }
```

Comment:
The logic sounds good, could we get it tested in test_deltafilewrapper.cpp ?

Reply:
I wanted to get it tested too, but I will not have the time to write the tests myself for now. Actually we have the two new functions from my recent PRs are also a very good candidates.